File: solutions/elixir/space-age/1/lib/space_age.ex
defmodule SpaceAge do
  @type planet ::
          :mercury
          | :venus
          | :earth
          | :mars
          | :jupiter
          | :saturn
          | :uranus
          | :neptune

  @doc """
  Return the number of years a person that has lived for 'seconds' seconds is
  aged on 'planet', or an error if 'planet' is not a planet.
  """
  @spec age_on(planet, pos_integer) :: {:ok, float} | {:error, String.t()}
  def age_on(planet, seconds) do
    case convert_age(planet) do
      :error -> {:error, "not a planet"}
      factor -> {:ok, seconds / (60 * 60 * 24 * 365.25) / factor}
    end
  end

  defp convert_age(planet) do
    case planet do
      :mercury -> 0.2408467
      :venus -> 0.61519726
      :earth -> 1.0
      :mars -> 1.8808158
      :jupiter -> 11.862615
      :saturn -> 29.447498
      :uranus -> 84.016846
      :neptune -> 164.79132
      _ -> :error
    end
  end
end

File: solutions/c/space-age/1/space_age.c
#include "space_age.h"

const float SECONDS_IN_YEAR = 31557600.0;

float age(planet_t planet, int64_t seconds)
{
  float orbital_period;
  switch (planet) {
    case MERCURY:
      orbital_period = 0.2408467;
      break;
    case VENUS:
      orbital_period = 0.61519726;
      break;
    case EARTH:
      orbital_period = 1.0;
      break;
    case MARS:
      orbital_period = 1.8808158;
      break;
    case JUPITER:
      orbital_period = 11.862615;
      break;
    case SATURN:
      orbital_period = 29.447498;
      break;
    case URANUS:
      orbital_period = 84.016846;
      break;
    case NEPTUNE:
      orbital_period = 164.79132;
      break;
    default:
      return -1.0;
  }

  return seconds / (SECONDS_IN_YEAR * orbital_period);
}
